### PR TITLE
feat: update old voice encryption modes (#1278)

### LINKED ISSUE #1278: Full Deprecated & Removal of Voice Connection Encryption Mode Discord 4J uses, beyond November 18th, 2024
As per

3.3.x: https://github.com/Discord4J/Discord4J/blob/master/voice/src/main/java/discord4j/voice/VoiceSocket.java#L54
3.2.x: https://github.com/Discord4J/Discord4J/blob/3.2.x/voice/src/main/java/discord4j/voice/VoiceSocket.java#L54

You can see Discord 4J is still using xsalsa20_poly1305_lite, this mode aswell as others are being fully deprecated to the point they will not be able to be used past November 18th, 2024

`The deprecated encryption modes above will be discontinued as of November 18th, 2024. As of this date the voice gateway will not allow you to connect with one of the deprecated encryption modes.`

6ee79b41415f7971d5bf73582e2a27e4

**Version:** All Branches, 3.2.x, 3.3.x

**Other:**

### PR DESCRIPTION
**Pull request by Azn9**
**Description:** Implement the new encryption mode `AEAD_AES256_GCM` and `AEAD_XCHACHA20_POLY1305`
**Justification:** Closes #1278, needed before November 18th, 2024

I tested both encryption modes, and they appear to be working as expected. At least I can connect a bot and have it play music; I didn't test receiving audio.

I admit I read JDA's code to find out why what I done was not working 😓

### CODE REVIEW

File: voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java [hunk 1]
```diff
@@ -260,10 +265,21 @@ private Mono<Void> connect(VoiceServerOptions vso, String sessionId,
                                                                 innerCleanup.add(connection);
                                                                 String hostName = address.getHostName();
                                                                 int port = address.getPort();
+
+                                                                this.encryptionMode = EncryptionMode.getBestMode();
+                                                                if (this.encryptionMode == null) {
+                                                                    log.error("No encryption mode available");
```
Comment (Doc94) [inline voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java:271]: this just not make show two messages of error?
one this log and latert the voiceConnectionSink?

File: voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java [hunk 2]
```diff
@@ -301,11 +330,14 @@ private Mono<Void> connect(VoiceServerOptions vso, String sessionId,
                                     .doOnNext(tick -> emissionStrategy.emitNext(outbound, tick))
                                     .then();
 
+                            String fullEndpoint = "wss://" + serverOptions.get().getEndpoint() + "?v=" + AUDIO_GATEWAY_VERSION; // TODO: support v8
```
Comment (Doc94) [inline voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java:333]: maybe this format can come from the VoiceServerOptions?

File: voice/src/main/java/discord4j/voice/PacketTransformer.java [hunk 3]
```diff
@@ -16,77 +16,148 @@
...
-        packet.getBytes(audioOffset, encrypted);
         packet.release();
 
-        byte[] decrypted = boxer.open(encrypted, getNonce(header));
-        if (decrypted == null) {
+        byte[] decrypted;
+        try {
+            decrypted = this.encryptionAdapter.decrypt(header, encrypted, nonce);
+        } catch (GeneralSecurityException e) {
+            PacketTransformer.LOGGER.debug("Failed to decrypt audio", e);
```
Comment (Doc94) [inline voice/src/main/java/discord4j/voice/PacketTransformer.java:140]: not sure if this is more for a warning rather than debug? or this not affect?

**Review by Doc94 — CHANGES_REQUESTED** [review]
Looks good, including the comments maybe can just take a look in the debug logs? maybe a little more info or if are necesary...

File: voice/src/main/java/discord4j/voice/VoiceServerOptions.java [hunk 4]
```diff
@@ -22,12 +22,14 @@
...
 
+    private static final int AUDIO_GATEWAY_VERSION = 4;
+
     private final String token;
     private final String endpoint;
 
     public VoiceServerOptions(String token, String endpoint) {
         this.token = token;
-        this.endpoint = endpoint.replace(":80", ""); // discord sends wrong port...
+        this.endpoint = "wss://" + endpoint + "?v=" + AUDIO_GATEWAY_VERSION;
```
Comment (Doc94) [inline voice/src/main/java/discord4j/voice/VoiceServerOptions.java:32]: the "replace" thing added its not needed anymore?

Comment (Azn9) [reply voice/src/main/java/discord4j/voice/VoiceServerOptions.java:32]: Yep, the port is now always :443

File: voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java [hunk 5]
```diff
@@ -238,11 +240,13 @@ private Mono<Void> connect(VoiceServerOptions vso, String sessionId,
                                     .flatMap(payloadReader)
                                     .doOnNext(payload -> {
                                         if (payload instanceof Hello) {
+                                            log.debug("Received HELLO");
```
Comment (Doc94) [inline voice/src/main/java/discord4j/voice/DefaultVoiceGatewayClient.java:243]: i dont feel this and the READY debug are necesary or just can me moved from the start of if with a trace for print the class if necesary.

**Pull request merged**